[quantization] Adding docstring to the observers

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27791 [quantization] Adding docstring to the observers**

This is the first part of the change. The next ones will amend more :)
Fixes #27849 

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

Commit:
[quantization] Adding docstring to the observers

This is the first part of the change. The next ones will amend more :)

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

z-a-f commented:
Fixes #27849

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -171,54 +197,88 @@ def _calculate_qparams(self, min_val, max_val):
+        if self.qscheme == torch.per_tensor_symmetric and \
+           self.reduce_range and \
+           self.dtype == torch.quint8:
+                raise NotImplementedError("Cannot reduce range for symmetric \
```
lint

jerryzh168 reviewed (approved):
LG, plz fix lint

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -9,44 +9,50 @@
 ABC = ABCMeta(str("ABC"), (object,), {})  # compatible with Python 2 *and* 3:
 
 
 class Observer(ABC, nn.Module):
```
Do we want to fix the name for Observer as well?

jerryzh168 replied:
I remember we want to rename this to `ObserverBase`

z-a-f replied:
That will be part of a different PR.

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -121,15 +142,20 @@ def _calculate_per_channel_qparams(self, min_vals, max_vals):
         # type: (Optional[Tensor], Optional[Tensor]) -> Tuple[Tensor, Tensor]
-        """
-        Given min and max values, this function calculates quantization parameters
+        """Calculates the per tensor quantization parameters, given the min/max.
```
why some docstring started with `r` some doesn't?

z-a-f replied:
Ideally we want all of them start with `r`. I will make sure they do.

The reason you want them to start with `r` is because when parsed by sphinx, the docstrings that contain `_` or `` ` `` will show some non-sense :)

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,86 @@ def _calculate_qparams(self, min_val, max_val):
+        x_\text{min/max} = \begin{cases}
+            \min(X) & \text{if~}x_\text{min/max} = \O \\
+            \min(x_\text{min/max}, \min(X)) & \text{otherwise}
+        \end{cases}
```
![image](https://user-images.githubusercontent.com/4216323/66855752-b24b5800-ef38-11e9-98dd-c413c0707dc6.png)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -238,10 +311,48 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        x_\text{min/max} = \begin{cases}
+            \min(X) & \text{if~}x_\text{min/max} = \O \\
+            (1 - c) x_\text{min/max} + c \min(X) & \text{otherwise}
+        \end{cases}
```
![image](https://user-images.githubusercontent.com/4216323/66855861-e7f04100-ef38-11e9-8011-72224f6e3616.png)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,86 @@ def _calculate_qparams(self, min_val, max_val):
+                &s = \left( x_\text{max} - x_\text{min}  \right ) /
+                    \left( Q_\text{max} - Q_\text{min} \right ) \\
+                &z = Q_\text{min} - \text{round}(x_\text{min} / s)
+        \end{aligned}
```
![image](https://user-images.githubusercontent.com/4216323/66856026-2a198280-ef39-11e9-828c-bb273ee0b430.png)

raghuramank100 replied:
For symmetric the equation should be 2.max(x_max, -x_min)/(Q_max -Q_min)

jerryzh168 replied:
use `Symmetric` and `Affine` to align the naming?

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,87 @@ def _calculate_qparams(self, min_val, max_val):
+            \frac{\min}{\max} & \text{if~}x_\text{min/max} = \O \\
+            \frac{\min}{\max}\left(x_\text{min/max},
+            \frac{\min}{\max}\left(X\right)\right) & \text{otherwise}
+        \end{cases}
```
![image](https://user-images.githubusercontent.com/4216323/66863643-4ec92680-ef48-11e9-86af-c5b124b15933.png)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -238,10 +312,48 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        x_\text{min/max} = \begin{cases}
+            \frac{\min}{\max} & \text{if~}x_\text{min/max} = \O \\
+            (1 - c) x_\text{min/max} + c \frac{\min}{\max} & \text{otherwise}
+        \end{cases}
```
![image](https://user-images.githubusercontent.com/4216323/66863667-5e486f80-ef48-11e9-8d70-0a3fee755230.png)

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,87 @@ def _calculate_qparams(self, min_val, max_val):
+            \frac{\min}{\max}(X) & \text{if~}x_\text{min/max} = \O \\
+            \frac{\min}{\max}\left(x_\text{min/max},
+            \frac{\min}{\max}(X)\right) & \text{otherwise}
+        \end{cases}
```
![image](https://user-images.githubusercontent.com/4216323/66863862-c5feba80-ef48-11e9-9980-1b772e0cedc2.png)

raghuramank100 replied:
Lets have separate equations for min and max, so that it is clearer

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -238,10 +312,48 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        x_\text{min/max} = \begin{cases}
+            \frac{\min}{\max}(X) & \text{if~}x_\text{min/max} = \O \\
+            (1 - c) x_\text{min/max} + c \frac{\min}{\max}(X) & \text{otherwise}
+        \end{cases}
```
![image](https://user-images.githubusercontent.com/4216323/66863915-e0389880-ef48-11e9-8eb0-4894805786c4.png)

raghuramank100 replied:
Same comments as before for min and max, Also, lets use None instead of phi

z-a-f reviewed (commented):
Added rendering of the math

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,88 @@ def _calculate_qparams(self, min_val, max_val):
+    where :math:`Q_\text{min}` and :math:`Q_\text{max}` are the minimum and
+    maximum of the quantized data type.
+
+    .. note:: Only works with ``torch.per_tensor_affine`` quantization shceme.
```
and per_tensor_symmetric quantization scheme

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -438,7 +603,7 @@ def _compute_quantization_error(self, next_start_bin, next_end_bin, norm_type):
 
     def _non_linear_param_search(self):
-        """
+        r"""
```
We should copy the comments from norm_minimization.cc here so that it is visible to end users

z-a-f replied:
I added the algo description in the class docstring

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,93 @@ def _calculate_qparams(self, min_val, max_val):
+            \max(X) & \text{if~}x_\text{max} = \text{None} \\
+            \max\left(x_\text{max}, \max(X)\right) & \text{otherwise}
+        \end{cases}\\
+        \end{array}
```
![image](https://user-images.githubusercontent.com/4216323/66872677-66f67100-ef5b-11e9-9f96-efb2ca3780b9.png)

jerryzh168 replied:
do we need this...

z-a-f replied:
I dunno -- maybe not. I think we can just say that this is a prior min and tensor min.

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -171,40 +197,93 @@ def _calculate_qparams(self, min_val, max_val):
+                &s = \left( x_\text{max} - x_\text{min}  \right ) /
+                    \left( Q_\text{max} - Q_\text{min} \right ) \\
+                &z = Q_\text{min} - \text{round}(x_\text{min} / s)
+        \end{aligned}
```
![image](https://user-images.githubusercontent.com/4216323/66872698-770e5080-ef5b-11e9-88e2-c423b61a2bc5.png)

z-a-f commented on torch/quantization/observer.py:
```diff
@@ -238,10 +318,54 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                    \max(X) & \text{if~}x_\text{max} = \text{None} \\
+                    (1 - c) x_\text{max} + c \max(X) & \text{otherwise}
+                \end{cases}\\
+        \end{array}
```
![image](https://user-images.githubusercontent.com/4216323/66872743-8a212080-ef5b-11e9-99b3-10646f71efa9.png)

jerryzh168 replied:
probably say `c` is `averaging_constant` here as well?

z-a-f replied:
> probably say `c` is `averaging_constant` here as well?

You mean in the equation? I say it in the text right after. So it will be rendered as equation followed by the variable description. I can add the description in the equation tho

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -64,13 +70,21 @@ def calculate_qparams(self, **kwargs):
+    custom observers.
+
+    Args:
+        dtype: Quantized data type
```
Can you also specify valid values for dtype and qscheme? For example., qint32 is not valid.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -350,25 +515,48 @@ def forward(self, x_orig):
+    1. Create the histogram of the incoming inputs.
+        The histogram is computed continuously, and the ranges per bin change
+        with every new tensor observed.
+    2. Search the distribution in the histogram for optimal min/max values.
```
Please update this comment

raghuramank100 reviewed (approved):
Looks good, thanks!

Commit:
Update on "[quantization] Adding docstring to the observers"


This is the first part of the change. The next ones will amend more :)
Fixes #27849 

Differential Revision: [D17889913](https://our.internmc.facebook.com/intern/diff/D17889913)
